### PR TITLE
Add a new webhook request total metric and add a 1.0s bucket boundary to the webhook latency metric

### PR DESCRIPTION
**Pull request by rmoriar1**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
We have a webhook which is configured to fail open and we would like to use metrics to detect if it is working properly. 

The existing `admission_webhook_rejection_count` metric doesn't track requests that failed open (because they are not considered rejections) and the `admission_webhook_admission_duration_seconds` only includes a bool indicating whether the request was rejected or not. As a result, we are unable to track the number of requests that have failed open.

This PR adds a new `admission_webhook_request_total` metric which includes the webhook's HTTP response  as a label. This will allow us to identify requests that have failed open. 

This also adds a new bucket boundary of 1.0s to the latency metric, which will enable us to get more accurate latency data from the webhook.

#### Does this PR introduce a user-facing change?
```release-note
Introduces a new metric: admission_webhook_request_total with the following labels: name (string) - the webhook name, type (string) - the admission type, operation (string) - the requested verb, code (int) - the HTTP status code, rejected (bool) - whether the request was rejected, namespace (string) - the namespace of the requested resource.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics_test.go [hunk 1]
```diff
@@ -85,15 +85,46 @@ func TestObserveAdmissionController(t *testing.T) {
 func TestObserveWebhook(t *testing.T) {
 	defer Metrics.reset()
 	defer legacyregistry.Reset()
-	Metrics.ObserveWebhook(context.TODO(), 2*time.Second, false, attr, stepAdmit, "x")
-	wantLabels := map[string]string{
+	Metrics.ObserveWebhook(context.TODO(), 2*time.Second, false, attr, stepAdmit, 200, "x")
```
Comment (alexcope) [inline staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics_test.go:88]: Generally I use context.Background() for tests

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics_test.go [hunk 2]
```diff
@@ -85,15 +85,46 @@ func TestObserveAdmissionController(t *testing.T) {
...
+		"namespace": "ns",
+		"code":      "654",
+	}
+	wantLabels000 := map[string]string{
+		"name":      "x",
+		"operation": string(admission.Create),
+		"type":      "validate",
 		"rejected":  "false",
 		"namespace": "ns",
+		"code":      "0",
```
Comment (alexcope) [inline staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics_test.go:122]: 0->200

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 3]
```diff
@@ -229,13 +231,13 @@ func (d *validatingDispatcher) callHook(ctx context.Context, h *v1.ValidatingWeb
...
 	if err := r.Do(ctx).Into(response); err != nil {
-		return &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: err}
+		return &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: fmt.Errorf("failed to call webhook: %w", err), Status: apierrors.NewBadRequest("Error executing webhook request")}
 	}
 	trace.Step("Request completed")
 
 	result, err := webhookrequest.VerifyAdmissionResponse(uid, false, response)
 	if err != nil {
-		return &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: err}
+		return &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: fmt.Errorf("received invalid webhook response: %w", err), Status: apierrors.NewBadRequest("error verifying webhook response")}
```
Comment (alexcope) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:240]: Not all of these are badRequest. I think this one should be NewClientErrorReporter. Read through the options here https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors?utm_source=godoc and pick the most appropriate one for each case

**Review by alexcope — CHANGES_REQUESTED** [review]
nevermind

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics_test.go (hunk 2 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics_test.go:122]: This case corresponds to the WebhookRejectionAPIServerInternalError case, which does return code 0 (i.e. no http code was set).

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go (hunk 3 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go]: Changed this one to a NewServiceUnavailable error. NewClientErrorReporter response is not ideal since that is of type ErrorReporter and we need it to be of type type StatusError.

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go [hunk 4]
```diff
@@ -136,7 +136,7 @@ func newAdmissionMetrics() *AdmissionMetrics {
 
 	// Admission webhook metrics. Each webhook is identified by name.
 	webhook := newMetricSet("webhook",
-		[]string{"name", "type", "operation", "rejected", "namespace"},
+		[]string{"name", "type", "operation", "rejected", "namespace", "code"},
```
Comment (dgrisonnet) [inline staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:139]: This will most likely make the cardinality explode for the histogram. Considering that the HTTP code is unbounded for admission webhooks, plus the fact that the histogram has 8 buckets, the number of series will theoretically increase by `8 * inf`. If we were to truncate codes to 600 as in `webhook_rejection_count`, it could theoretically increase by `8 * 600` which is a lot. Even though this shouldn't be the case in practice, it is still worrying.

For these reasons, the HTTP code isn't usually added to the latency histogram, but rather to the request counter. Thus, an idea could be to replace `webhook_rejection_count` by a `webhook_request_total` metric that would account for all HTTP codes. We could also truncate the HTTP code to 600 to bound the cardinality of the metric.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 5]
```diff
@@ -229,13 +231,13 @@ func (d *validatingDispatcher) callHook(ctx context.Context, h *v1.ValidatingWeb
 	}
 
 	if err := r.Do(ctx).Into(response); err != nil {
-		return &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: err}
+		return &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: fmt.Errorf("failed to call webhook: %w", err), Status: apierrors.NewBadRequest("error calling webhook")}
```
Comment (rmoriar1) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:234]: If r.Do() returns a StatusError we should propagate it instead of always returning NewBadRequest.

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go (hunk 4 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:139]: Good point.

I agree that `webhook_request_total` would be a more useful metric if we were starting from scratch, but for backward compatibility purposed it's probably best to leave `webhook_rejection_count` as is and instead bound the cardinality of the response codes.

Including the code in the latency metric can be useful for identifying if latency spikes are correlated to a specific error.

Comment (dgrisonnet) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:139]: Even if we were to introduce `webhook_request_total` now, it would generate fewer series than adding the HTTP code to the histogram.

`webhook_rejection_count` is still an ALPHA metric, so we can change it anytime. I am not too familiar with admission webhooks, but shouldn't `webhook_request_total` covers the purpose of having `webhook_rejection_count` and handle your fail open use case at the same time, or are rejections handled differently and can't be deduced from the HTTP code only?

I would agree with the statement that the code can be useful in latency metrics, but I don't think it is worth the cardinality spike in this particular use case, but again I am not too familiar with admission webhooks.
Generally, requests failing with code 4xx or 5xx are faster since they are rejected sooner in the normal code path, so if we only care about the 90th and 99th percentile of request latency, for example, to determine SLOs, then the code would not be really useful. Moreover, when determining SLOs and creating alerts on burn rate, what we really care about are the requests with the highest latency (99th percentile) and the ratio of error requests which can be captured by a `request_total` metric.

Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:139]: `webhook_rejection_count` is already being used in a number of places internally at google, so I think modifying it is a non-starter. webhook rejections can be deduced from the HTTP code since they are [converted to >=400](https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/errors/statuserror.go#L37), so I'm not entirely sure why the rejection count metric was chosen in the first place.

That's a good point about failing requests being unlikely to contribute to tail latencies. If the cardinality spike is not worth it, I think the best course of action is to add a new `webhook_request_total` metric leaving the other two metrics as is.

I'll update the PR with the changes hopefully in the next day or two.

Comment (dgrisonnet) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: @logicalhan mentioned in yesterday's sig-instrumentation triage meeting that the cardinality spike might not be too problematic since we have some mechanisms to prevent cardinality explosions in k/k. That said, I am still on the fence regarding adding code to the histogram since it might not be very useful to the latency, but happy to discuss it :)

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go [hunk 6]
```diff
@@ -149,11 +150,22 @@ func newAdmissionMetrics() *AdmissionMetrics {
 		},
 		[]string{"name", "type", "operation", "error_type", "rejection_code", "namespace"})
 
+	webhookRequest := metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "webhook_request_count",
```
Comment (dgrisonnet) [inline staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:157]: According to [prometheus best practices](https://prometheus.io/docs/practices/naming/), counters should be suffixed by `_total`.
```suggestion
			Name:           "webhook_request_total",
```

Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go (hunk 4 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: @logicalhan seemed to be of the mind that adding the new metric was best when I chatted with him. Since that fits our use case just fine I don't think there's much of a need to modify the latency metric at present.

Comment (dgrisonnet) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: :+1:. Would it make sense to work on the deprecation of `webhook_rejection_count` in favor of `webhook_request_total` in a follow-up effort?

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go [hunk 7]
```diff
@@ -173,7 +185,12 @@ func (m *AdmissionMetrics) ObserveAdmissionController(ctx context.Context, elaps
 }
 
 // ObserveWebhook records admission related metrics for a admission webhook.
-func (m *AdmissionMetrics) ObserveWebhook(ctx context.Context, elapsed time.Duration, rejected bool, attr admission.Attributes, stepType string, extraLabels ...string) {
+func (m *AdmissionMetrics) ObserveWebhook(ctx context.Context, elapsed time.Duration, rejected bool, attr admission.Attributes, stepType string, code int, extraLabels ...string) {
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:188]: what's going on with this method signature? it takes var args but all usages only pass in a single argument. from what i can tell, if there were more than one argument passed in the label values would be misaligned in line 194 (or possibly overflow).

Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: Yes, this was confusing to me as well. Converted this to a non-variadic function and added the name of the webhook as an arg, similar to `ObserveWebhookRejection`.

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: awesome, thanks for doing that!

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go (hunk 4 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: This would require a fair amount of work internally. @logicalhan has more context here so I'll defer the decision to him.

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve
(from the metrics side)

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go [hunk 8]
```diff
@@ -149,11 +150,22 @@ func newAdmissionMetrics() *AdmissionMetrics {
...
 
+	webhookRequest := metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "webhook_request_total",
+			Help:           "Admission webhook request total, identified by name and broken out for each admission type (validating or admit) and operation. Additional labels specify an HTTP status code. Codes greater than 600 are truncated to 600, to keep the metrics cardinality bounded.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"name", "type", "operation", "code", "namespace"})
```
Comment (rmoriar1) [inline staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:161]: On second thought, I think a rejection bool is needed here in order to distinguish between a rejected request with a certain code and a failed open request with the same code.
This could be done in a roundabout way by subtracting away rejections using the `rejection_count` metric, but that seems less than ideal.

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: i'd add it now then, this is a candidate metric we'd want to promote to stable and if so we wouldn't want to make any breaking changes.

**Review by logicalhan — COMMENTED** [review]
This does contain a user facing change so you should amend the release note with the new metric that this introduces.

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go (hunk 8 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: Done.

**Review by logicalhan — COMMENTED** [review]
/lgtm

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go [hunk 9]
```diff
@@ -149,11 +150,22 @@ func newAdmissionMetrics() *AdmissionMetrics {
 		},
 		[]string{"name", "type", "operation", "error_type", "rejection_code", "namespace"})
 
+	webhookRequest := metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "webhook_request_total",
+			Help:           "Admission webhook request total, identified by name and broken out for each admission type (validating or admit) and operation. Additional labels specify an HTTP status code. Codes greater than 600 are truncated to 600, to keep the metrics cardinality bounded.",
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:158]:
```suggestion
			Help:           "Admission webhook request total, identified by name and broken out for each admission type (validating or mutating) and operation. Additional labels specify an HTTP status code. Codes greater than 600 are truncated to 600, to keep the metrics cardinality bounded.",
```

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go [hunk 10]
```diff
@@ -149,11 +150,22 @@ func newAdmissionMetrics() *AdmissionMetrics {
...
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"name", "type", "operation", "code", "rejected", "namespace"})
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:161]: What if we switched rejected for permitted? It's easier for humans to reason about things stated in the positive?

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go [hunk 11]
```diff
@@ -173,12 +185,17 @@ func (m *AdmissionMetrics) ObserveAdmissionController(ctx context.Context, elaps
...
+	if code > 600 {
+		code = 600
+	}
+	m.webhookRequest.WithContext(ctx).WithLabelValues(name, stepType, string(attr.GetOperation()), strconv.Itoa(code), strconv.FormatBool(rejected), attr.GetNamespace()).Inc()
+	m.webhook.observe(ctx, elapsed, name, stepType, string(attr.GetOperation()), strconv.FormatBool(rejected), attr.GetNamespace())
 }
 
 // ObserveWebhookRejection records admission related metrics for an admission webhook rejection.
-func (m *AdmissionMetrics) ObserveWebhookRejection(ctx context.Context, name, stepType string, attr admission.Attributes, errorType WebhookRejectionErrorType, rejectionCode int) {
+func (m *AdmissionMetrics) ObserveWebhookRejection(ctx context.Context, name string, stepType string, attr admission.Attributes, errorType WebhookRejectionErrorType, rejectionCode int) {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:198]: revert unrelated change?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 12]
```diff
@@ -261,13 +265,18 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *admiss
 	}
 
 	if err := r.Do(ctx).Into(response); err != nil {
-		return false, &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: fmt.Errorf("failed to call webhook: %w", err)}
+		switch err := err.(type) {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:268]: This makes it easier to see what is different about the two cases:

```
var status 
if se, ok := err.(*apierrors.StatusError); ok {
  status = se
} else {
  status = apierrors.NewBadRequest("error calling webhook")
}
return false, &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: fmt.Errorf("failed to call webhook: %w", err), Status: status}
```

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 13]
```diff
@@ -229,13 +231,18 @@ func (d *validatingDispatcher) callHook(ctx context.Context, h *v1.ValidatingWeb
 	}
 
 	if err := r.Do(ctx).Into(response); err != nil {
-		return &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: err}
+		switch err := err.(type) {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:234]: same suggestion here.

**Comment by lavalamp** [comment]
looks generally good, some suggestions though.

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go (hunk 9 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go (hunk 10 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: sgtm, the only drawback is the latency metric uses rejected.

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go (hunk 11 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: Sure, but what's the rationale? To keep the PR as narrow as possible?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go (hunk 12 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: Nice!

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go (hunk 13 above)
Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go (hunk 11 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: Right.

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go (hunk 10 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: Oh, I'm sorry, I looked for other metrics and didn't see that one while skimming, only the rejection one, and we should probably be consistent, sigh.

Uh, but now that you point it out, the only thing this counter metric is adding is the response code. Why do we need an entirely new metric just for that? You should be able to get a count out of a histogram, no?

Comment (rmoriar1) [reply staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: Okay, I'll revert back to using rejected.

There is a comment thread above this discussing whether the code should be added to the latency metric or as an entirely new metric. The downside of adding this to the latency metric is it would make the cardinality explode for the histogram.

**Comment by rmoriar1** [comment]
> @rmoriar1 needs rebase

Done.

**Comment by rmoriar1** [comment]
@lavalamp Mind re-reviewing after rebase?
